[llvm][cas] Add validate-if-needed to recover from invalid data

benlangmuir opened the pull request:
Introduce a new validate-if-needed API to the UnifiedOnDiskCache and
llvm-cas tool that triggers out-of-process validation of the CAS once
for every machine boot, and optionally recovers from invalid data by
marking it for garbage collection. This fixes a hole in the CAS data
coherence when a power loss or similar failure causes the OS to not
flush all of the pages in the mmaped on-disk CAS files.

The intent is that clients such as the clang scanning daemon or a build
system should trigger this validation at least once before using the
CAS.

rdar://123542312

benlangmuir committed:
[llvm][cas] Add validate-if-needed to recover from invalid data

Introduce a new validate-if-needed API to the UnifiedOnDiskCache and
llvm-cas tool that triggers out-of-process validation of the CAS once
for every machine boot, and optionally recovers from invalid data by
marking it for garbage collection. This fixes a hole in the CAS data
coherence when a power loss or similar failure causes the OS to not
flush all of the pages in the mmaped on-disk CAS files.

The intent is that clients such as the clang scanning daemon or a build
system should trigger this validation at least once before using the
CAS.

rdar://123542312

benlangmuir committed:
[clang][cas] Adopt validate-if-needed in clang scanning daemon

Use the new validate-if-needed functionality to ensure the clang
scanning daemon's CAS data is valid.

benlangmuir commented:
@swift-ci please test llvm

cachemeifyoucan commented on llvm/lib/CAS/UnifiedOnDiskCache.cpp:
```diff
@@ -194,6 +249,230 @@ static void getNextDBDirName(StringRef DBDir, llvm::raw_ostream &OS) {
+                                  bool CheckHash) {
+  SmallVector<StringRef> Args{LLVMCasBinary, "-cas", RootPath, "-validate"};
+  if (CheckHash)
+    Args.push_back("-check-hash");
```
We need to get the plugin options in there somehow.

benlangmuir replied:
This is the on-disk version, so it doesn't use the plugin by definition.

cachemeifyoucan commented on llvm/lib/CAS/UnifiedOnDiskCache.cpp:
```diff
@@ -194,6 +249,230 @@ static void getNextDBDirName(StringRef DBDir, llvm::raw_ostream &OS) {
+    sys::path::remove_filename(PathBuf);
+    sys::path::append(PathBuf, "lock");
+
+    int LockFD = -1;
```
I wonder if it is a good idea to lock before validate (so there is no need to acquire lock here). At least that prevents CAS being broken as the validation is run.

Also maybe the recovery should be done out of process if needed as well instead of always perform in-process?

benlangmuir replied:
> Also maybe the recovery should be done out of process if needed as well instead of always perform in-process?

What is the benefit of recovering out of process?

benlangmuir replied:
> I wonder if it is a good idea to lock before validate (so there is no need to acquire lock here).

I considered this, but there are two main advantages of only locking when we need to modify the CAS data:
* It makes this implementation more lenient when working with tools that don't have validation yet or that don't call validation thoroughly, because it will only cause an error if there is actually invalid data. If we locked exclusive immediately then we either need to block indefinitely for other cas users or you get an error even if there is no invalid data.
* It simplifies the locking because the out-of-process validation can just use the shared lock like normal without needing a new code path.

For
> At least that prevents CAS being broken as the validation is run.

For once-per-boot validation to protect against missing flushed pages, it is outside the use-case to care about the CAS being made invalid by our own tooling or mid-way through operation. We could consider changing this down the road if we do period revalidation to catch bugs, but I prefer the current version especially as we are first rolling it out and likely to have tools that don't call it operating from the same CAS.

cachemeifyoucan replied:
I assume that since `llvm-cas` can recover, maybe just let it do the work. I am fine either way and we can make future adjustments if needed.

benlangmuir replied:
I mentioned this offline already, but to make sure it's captured:
The reason we don't do recovery in the spawned sub process is that we also want recovery to work if validation crashes. If we can improve validation so it is safe enough to run in-process, that would be ideal in the future.

cachemeifyoucan commented on llvm/include/llvm/CAS/BuiltinUnifiedCASDatabases.h:
```diff
@@ -21,6 +21,39 @@ class ObjectStore;
+  /// The data is already valid.
+  Valid,
+  /// The data was invalid, but was recovered.
+  RecoveredValid,
```
Maybe just call this `Recovered`?

cachemeifyoucan reviewed (commented):
Most of the implementation is fine to me. Comments inline.

I think we need to add the plugin version (at least a placeholder) so the APIs can be settled.

benlangmuir commented:
> I think we need to add the plugin version (at least a placeholder) so the APIs can be settled.

I want to do this as a follow-up. The API will look basically the same except with the plugin path and options, but it's totally separate in theory since it's a new top-level API not just a method you call on an existing CAS instance.

cachemeifyoucan commented on clang/lib/CAS/CASOptions.cpp:
```diff
@@ -119,3 +119,12 @@ llvm::Error CASOptions::initCache() const {
   return llvm::Error::success();
 }
+
+std::string CASOptions::getResolvedCASPath() const {
```
I am slightly prefer to have this in the tool if this function only has one call site. I think we want to sunset the `auto` name in general.

benlangmuir replied:
I'm fine with getting rid of auto in the future, but I need this API to match `CASOptions::initCache()` since it's still in use.  I changed it so both use this API to keep it in sync.

cachemeifyoucan reviewed (commented):
I see. The pluginCAS doesn't actually share the APIs here. We can do that as followup.

benlangmuir committed:
Updates for review

benlangmuir committed:
Handle EEXIST for Linux rename failure

Documentation says it can either be ENOTEMPTY (like Darwin) or EEXIST.
Also print the error.

benlangmuir commented:
@swift-ci please test llvm
